### PR TITLE
Replay sequence detection

### PR DESCRIPTION
**Pull request by josefbitzenhofer**
## Summary by CodeRabbit

* **New Features**
  * Full offline sequence detection and Bayesian decoding pipeline with Radon-based trajectory analysis, visualization, diagnostics, and en‑bloc vs per‑event modes; plotting and summary statistics exports.

* **New Models**
  * Structured datatypes to standardize decoding inputs/outputs and replay/LUT results.

* **Tests**
  * Extensive unit tests for event detection, merging, duration filtering, spatial binning, ensemble extraction, and threshold utilities.

* **Bug Fixes**
  * Runtime safeguards for legacy preprocessing and imaging edge-case handling.

* **Refactor**
  * Simplified imaging and ensemble APIs; added speed/occupancy filtering and enhanced utilities.

* **Chores**
  * Session mapping, caching/progress feedback, .gitignore expanded, and dependency manifest updated.

### CODE REVIEW

File: viral/offline_sequence_detection.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+        # look for end (go forwards until below 1 SD)
+        end_idx = peak_idx
+        while (
+            end_idx < len(population_vector) - 1
+            and population_vector[end_idx]
+            > config.edge_threshold * population_vector_sd
+        ):
+            end_idx += 1
+        events.append((start_idx, end_idx))
+
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py]: Well, it currently stops right at the edge, correct? I think this could be ok

File: requirements.txt [hunk 2]
```diff
@@ -1,88 +1,76 @@
...
 pandas==2.2.3
-parso==0.8.4
-pexpect==4.9.0
+patsy==1.0.2
 pillow==11.1.0
-platformdirs==4.3.6
 pluggy==1.5.0
-prompt_toolkit==3.0.50
 proto-plus==1.26.0
 protobuf==5.29.3
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply requirements.txt:56]: You're right and I should also not get rid of most of the deleted requirements just because I hadn't installed them on my machien

File: viral/sequence_utils.py [hunk 3]
```diff
@@ -0,0 +1,924 @@
...
+        last_start, last_end = merged_events[-1]
+        if current_start - last_end < 6:  # less than 0.2s apart
+            merged_events[-1] = (
+                last_start,
+                current_end,
+            )  # merge two events that are too close to each other
+        else:
+            # not too close, keep the event
+            merged_events.append((current_start, current_end))
+    return merged_events
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/sequence_utils.py:90]: In offline_sequence_detection.py, the function hits an early return if the input to this function is an empty list, so no need to do it here

File: tests/test_ensembles.py [hunk 4]
```diff
@@ -80,3 +84,396 @@ def check_any_shuffles_the_same(arr: np.ndarray) -> bool:
...
+    mock_trials_all_chunks_filtered_out,
+) -> None:
+    trials, place_cells = mock_trials_all_chunks_filtered_out
+    sigma = 30
+    mode = "above"
+    speed_threshold = 1
+    n_consecutive_samples = 1
+    min_chunk_len = 4
+
+    with patch(
```
Comment (neuromantic99) [inline tests/test_ensembles.py:358]: I might have misinterpreted the name of this test, but shouldn't this filter out all the flurosence if it's called "all chunks filtered out"?

File: viral/ensemble_reactivation.py [hunk 5]
```diff
@@ -660,7 +659,12 @@ def plot_grosmark_panel(
 def get_ssp_vectors(
```
Comment (neuromantic99) [inline viral/ensemble_reactivation.py:663]: wow that looked painful to get out when the chunks started, good job

File: viral/models.py [hunk 6]
```diff
@@ -210,3 +210,88 @@ def __post_init__(self):
...
+
+@dataclass
+class SSPConfig:
+    mode: Literal["below", "above"]
+    speed_threshold: float
+    n_consecutive_samples: int
+
+
+@dataclass
+class SSPVectorData:
```
Comment (neuromantic99) [inline viral/models.py:224]: can you clarify the chunk_start_indices comment a bit and add a comment for the other variables, it took me a while to unpick

File: viral/grosmark_analysis.py [hunk 7]
```diff
@@ -97,8 +101,11 @@ def get_place_cells(
     spks: np.ndarray,
     config: GrosmarkConfig,
     rewarded: bool | None,
+    use_cache: bool = True,
+    bin_occupancy_divide: bool = False,
     plot: bool = True,
-) -> tuple[np.ndarray, np.ndarray]:
+    use_train_test_split: bool = False,
```
Comment (neuromantic99) [inline viral/grosmark_analysis.py]: Something slightly odd about using a boolean flag just to change the cache path that you save to. When I first read it, I thought this function now had test-train split functionality but it's just to rename the cache file right? Maybe it should be an argument that's like "cache_file_additional_info: str | None = None and you can add it in like that

File: viral/offline_sequence_detection.py [hunk 8]
```diff
@@ -0,0 +1,1579 @@
...
+        trials_train, trials_test = train_test_split(
+            trials, train_size=train_size, random_state=42
+        )
+        use_train_test_split = True
+
+        # all trial frames
+        # ssp_test, positions_test = get_ssp_vectors(trials_test, place_cells, bayesian_config.sigma_online, "all", 5, 3 * 30)
+
+        # TODO: change back!
+        # phases of mobility
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]: 😬

File: viral/offline_sequence_detection.py [hunk 9]
```diff
@@ -0,0 +1,1579 @@
...
+    plt.close()
+
+
+def main(
+    mouse_name: str,
+    date: str,
+    bayesian_config: BayesianDecodingConfig,
+    grosmark_config: GrosmarkConfig,
+) -> BayesianDecodingResult | None:
+    """
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]: Unfortunately I can't follow the logic of this function. The main issue is the use of lots of nested if/else statements with very long blocks below.

I would refactor this, the lowest hanging fruit is to return out of the function when you can, for example:

Do this:
```python
def func():
    if foo:
        return 1
    return 2
```

Not this:
```python
def func():
    if foo:
        x = 1
    else:
        x = 2
    return x
```

This should reduce the number of nestings

Youll also probably need to break out some of the logic into separate functions. E.g.

```python
def main()
    if enc_block:
        return process_enc_block()
```

rather than

```python
def main()
    if enc_block:
        .... loads of stuff ....
    else:
        ... loads of stuff ...
return result
```

File: viral/offline_sequence_detection.py [hunk 10]
```diff
@@ -0,0 +1,1579 @@
...
+                )
+            # TODO: is this sigma correct?
+            ssp_test = gaussian_filter1d(
+                input=offline,
+                sigma=bayesian_config.sigma_offline,
+                axis=1,
+            )
+
+        if not bayesian_config.en_bloc:
+            if not bayesian_config.epoch == "online":
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]:
```suggestion
            if bayesian_config.epoch != "online":
```

File: viral/models.py [hunk 11]
```diff
@@ -210,3 +210,88 @@ def __post_init__(self):
...
+class BayesianDecodingConfig:
+    peak_threshold: float  # SDs above mean
+    edge_threshold: float  # SDs above mean
+    event_duration: Tuple[float, float]  # min, max (frames)
+    bin_size_time_offline: int  # frames
+    bin_size_time_online: int  # frames
+    start_spatial: int  # cms
+    end_spatial: int  # cms
+    bin_size_spatial: int  # cms
+    en_bloc: bool  # whether to decode en bloc or pse-event-by-pse-event/trial-chunk-by-trial-chunk
```
Comment (neuromantic99) [inline viral/models.py]: what's an en block?

Sounds like a french term

File: viral/models.py [hunk 12]
```diff
@@ -210,3 +210,88 @@ def __post_init__(self):
...
+    epoch: Literal["pre", "online", "post"]
```
Comment (neuromantic99) [inline viral/models.py:247]: is this online: online in the sense that it's when the mouse is running? Or decoding when the mouse rested during the task?

File: viral/offline_sequence_detection.py [hunk 13]
```diff
@@ -0,0 +1,1579 @@
...
+
+def get_population_vector(
+    ssp_smoothed: np.ndarray, mouse_name: str, date: str
+) -> np.ndarray:
+    """
+    Offline PSEs were detected by convolving each PC's (as assessed during that day's run) offline immobility firing rate vector
+    Ssp with a 125-ms Gaussian kernel and z-scoring the smoothed firing rate vector. Subsequently, for each frame i, the population
+    mean of the smoothed and z-scored vector was taken across PCs and subsequently z-scored.
+    """
+    # TODO: check axis?
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]: that's correct

File: viral/offline_sequence_detection.py [hunk 14]
```diff
@@ -0,0 +1,1579 @@
...
+    z_scored = zscore(ssp_smoothed, axis=1)
+    # Remove nans from silent neurons
+    z_scored = np.nan_to_num(z_scored)
+
+    # TODO: check axis?
+    # return zscore(np.mean(z_scored, axis=0))
+    population_vector = zscore(np.mean(z_scored, axis=0))
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py:69]: The double z-scoring seems weird to me but that is what it says in the paper

File: viral/offline_sequence_detection.py [hunk 15]
```diff
@@ -0,0 +1,1579 @@
...
+        xmax=len(population_vector),
+        colors="g",
+        linestyles="dashed",
+        label="Edge threshold",
+    )
+    plt.title("Population vector")
+    plt.savefig(
+        PLOT_PATH
+        / f"{'pse_events_online' if bayesian_config.online else 'pse_events_offline'}"
+        / f"{mouse_name}_{date}_{'online' if bayesian_config.online else 'offline'}_population_vector.png",
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]:
```suggestion
        / f"{'pse_events_online' if bayesian_config.epoch == 'online' else 'pse_events_offline'}"
        / f"{mouse_name}_{date}_{'online' if bayesian_config.epoch == 'online' else 'offline'}_population_vector.png",
```

File: viral/offline_sequence_detection.py [hunk 16]
```diff
@@ -0,0 +1,1579 @@
...
+    duration_filter: bool = True,
+) -> List[Tuple[int, int]]:
+    """
+    Putative PSEs were defined as epochs during which the z-scored population activity vector reached a peak of at least 3.5 s.d.
+    above the mean with event-edges at 1 s.d. above the mean, with a minimum inter-event time of 0.2 s.
+    Only PSE events lasting between 0.2 s (12 frames) and 1 s (60 frames), and during which at least 5 distinct PCs each fired at
+    least one estimated spike, were kept for further analysis.
+    """
+
+    # TODO: is the z_scored population activity vector the one with the z_scored means????
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]: i think you've done this correctly

File: viral/offline_sequence_detection.py [hunk 17]
```diff
@@ -0,0 +1,1579 @@
...
+    bayesian_config = BayesianDecodingConfig(
+        en_bloc=False,
+        epoch="online",
+        # TODO: or is the 125 ms sigma also just offline and 1 s for online????
+        sigma_offline=(125 / 1000) * 30,  # "125 ms Gaussian kernel"
+        sigma_online=30,  # "1 s Gaussian kernel"
+        peak_threshold=3.5,
+        edge_threshold=1,
+        # event_duration=(6, 30),
+        event_duration=(6, 120),
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py:1506]: do you think 4 seconds for the upper bound may be slightly too long?

File: viral/offline_sequence_detection.py [hunk 18]
```diff
@@ -0,0 +1,1579 @@
...
+            event_duration_thresholds=config.event_duration,
+        )
+        print(f"Filtered to {len(filtered_events)} events by duration")
+        if len(filtered_events) == 0:
+            return filtered_events
+    else:
+        filtered_events = merged_events
+        print("No duration filtering at the moment")
+
+    # TODO: ssp is estimated spikes, right?
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py:138]: Yes, I think technically this should be the unsmoothed version as you mayyyy get edge effects but this would be such a pain to do i wouldn't bother

File: viral/sequence_utils.py [hunk 19]
```diff
@@ -0,0 +1,954 @@
...
+            filtered_events.append((start_idx, end_idx))
+    return filtered_events
+
+
+def additional_pc_check(
+    filtered_events: List[Tuple[int, int]], ssp: np.ndarray
+) -> List[Tuple[int, int]]:
+    additionally_checked = list()
+    for start_idx, end_idx in filtered_events:
+        pcs_with_spikes = np.where(np.sum(ssp[:, start_idx:end_idx], axis=1) >= 1)[0]
```
Comment (neuromantic99) [inline viral/sequence_utils.py:114]: I think the way you've done this is fine (summing across the bins). I think the more correct way as I said in previous comment would be to look at the unsmoothed. In the smoothed data, you will need to do something like summing (as you've done) because the actual value of a smoothed and sparisified binary vector will probably never >= 1.

I think with your summing approach you could possibly reject cells that have actually spiked. As if they spike once in the PSE, the smoothing will reduce this value massively (if it doesn't spike at all 1 second either side of the single spike, the value will be something like (1/30) (not exactly but not far off).

File: viral/offline_sequence_detection.py [hunk 20]
```diff
@@ -0,0 +1,1579 @@
...
+                    config=bayesian_config,
+                )
+
+                if len(pse_events) == 0:
+                    print("No PSE events found, exiting")
+                    return
+
+                pse_activity = [ssp_test[:, start:end] for start, end in pse_events]
+
+                # TODO: why can place_fields contain NaNs???
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py]: if you're speed thresholding it may do, if you're not, it shouldn't

File: viral/offline_sequence_detection.py [hunk 21]
```diff
@@ -0,0 +1,1579 @@
...
+
+    if np.isinf(Pr).any():
+        raise ValueError("Infinite values in Pr")
+    if np.isnan(Pr).any():
+        raise ValueError("NaN values in Pr")
+
+    # and only bins with non-zero firing rates were used for offline Bayesian decoding. Posterior probabilities were subsequently normalized to one:
+
+    return Pr, pr_max
+    # TODO: check the matlab implementation for same result
```
Comment (neuromantic99) [inline viral/offline_sequence_detection.py:225]: There's no point me reviewing this direclty i don't think, I will check myself against matlab :)

File: viral/sequence_utils.py [hunk 22]
```diff
@@ -0,0 +1,954 @@
...
+    )
+    rcs = calculate_linear_weighted_correlation(
+        posterior_probability_matrix, np.column_stack((np.sin(y), np.cos(y)))
+    )
+
+    return np.sqrt((rxc**2 + rxs**2 - 2 * rxc * rxs * rcs) / (1 - rcs**2))
+
+    # TODO: check against Matlab
+    # TODO: for perfect diagonal?
+    # TODO: for slightly off-diagonal: Python 0.7583601138637951 MATLAB 0.7584 (set pi to 3.14 to prevent pi-precision errors) -> AssertionError -> but fine? what tolerance?
```
Comment (neuromantic99) [inline viral/sequence_utils.py:268]: again checking against matlab is better than reviewing

File: viral/sequence_utils.py [hunk 23]
```diff
@@ -0,0 +1,954 @@
...
+    """
+    "For each event, the observed weighted circo-linear correlation coefficients weightedr(circular), hereafter referred to as weighted-r,
+    was compared to a distribution of 2,000 null weighted-r values either by z-scoring the observed value by the null values (rZ score) or
+    as the empirical P value.
+    While several shuffle approaches were used (Extended Data Fig. 6),
+    the principal shuffle used in the main figures involved the random re-ordering (resampling without replacement)
+    of the bins observed within a given event."
+    "'timeBinPermutation': permutes (resamples without replacement)"
+
+    Essentially, this is a Python implementation of https://github.com/losonczylab/Grosmark_NatNeuro_2021/blob/main/shufflePopulationEvents.m
```
Comment (neuromantic99) [inline viral/sequence_utils.py:290]: have you checked that this is the same as matlab        ?

File: viral/sequence_utils.py [hunk 24]
```diff
@@ -0,0 +1,954 @@
...
+    mode: Literal["linear", "circular"],
+    total_length: Optional[float] = None,
+) -> np.ndarray:
+    """
+    We need to precompute a xy-by-bin grid for linear/circular weighted correlation.
+
+    Essentially, this is a Python implementation of the section in
+    https://github.com/losonczylab/Grosmark_NatNeuro_2021/blob/main/demo_LinearReplayAnalysis.m
+    and https://github.com/losonczylab/Grosmark_NatNeuro_2021/blob/main/demo_CircularReplayAnalysis.m, respectively.
+    """
```
Comment (neuromantic99) [inline viral/sequence_utils.py:131]: does it work the same?

File: viral/sequence_utils.py [hunk 25]
```diff
@@ -0,0 +1,954 @@
...
+    and pol2cart.m (MATLAB).
+    """
+    # x = r.*cos(th);
+    # y = r.*sin(th);
+    x = rho * np.cos(phi)
+    y = rho * np.sin(phi)
+    return (x, y)
+
+
+# TODO: how could we confirm this is correct??
```
Comment (neuromantic99) [inline viral/sequence_utils.py:346]: I guess is the function that calls this is the same in matlab and python then you're good

File: viral/models.py (hunk 12 above)
Comment (josefbitzenhofer) [reply viral/models.py:247]: Online in the sense that the mouse is running

File: viral/models.py (hunk 11 above)
Comment (josefbitzenhofer) [reply viral/models.py]: Originally, I couldn't detect any meaningful events back in November. After we had a chat, I was looking at the entire epoch (I think the post wheel freeze) as a whole. In surgery, when you remove an organ as a continuous piece with healthy tissue surrounding it instead of cutting it into smaller pieces, you call that procedure an "en bloc resection" (at least in German, that is). I used this switch as a placeholder to decode the entire epoch but I think I can remove it now as we have seen events in the post as well as in the online and couldn't really use this "en bloc" at all.
Looked it up in the dictionary and en bloc means "all together or all at the same time".

File: viral/ensemble_reactivation.py (hunk 5 above)
Comment (josefbitzenhofer) [reply viral/ensemble_reactivation.py:663]: Was a lot of back and forth, but I'll take the compliment ^^

File: viral/models.py (hunk 6 above)
Comment (josefbitzenhofer) [reply viral/models.py:224]: Sorry yeah, should've explained the nested list 😅

File: viral/offline_sequence_detection.py (hunk 18 above)
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py:138]: So just leave it as it is?

File: viral/offline_sequence_detection.py (hunk 14 above)
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py:69]: yepppp

File: viral/offline_sequence_detection.py (hunk 8 above)
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py]: nooooo, you don't understand.....
I was going back and forth and couldn't remember what the last thing was we wanted to look at (mobility/immobility), you have to believe me I didn't just overlook my own comment....

File: viral/offline_sequence_detection.py (hunk 17 above)
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py:1506]: Yep it is too long. In another PR I changed it to 90 (i.e. 3 seconds), but probably we want more like 75 or 60 (i.e. 2.5 or 2 seconds) - or even just 45 (i.e. 1.5 seconds)?

File: viral/offline_sequence_detection.py (hunk 20 above)
Comment (josefbitzenhofer) [reply viral/offline_sequence_detection.py]: Ok, yeah that's what I thought. Should disappear when I start using the BOD instead.

File: tests/test_ensembles.py (hunk 4 above)
Comment (josefbitzenhofer) [reply tests/test_ensembles.py:358]: I should have given it a more descriptive name or added a comment.
It is supposed to filter out all chunks in the FIRST trial only, as they are too short. Added a comment now.

File: viral/grosmark_analysis.py [hunk 26]
```diff
@@ -66,7 +70,7 @@ def grosmark_place_field(
         spks = np.hstack([offline_spks_pre, online_spks, offline_spks_post])
         assert spks_raw.shape == spks.shape
 
-    pcs, smoothed_matrix = get_place_cells(
+    pcs, smoothed_matrix, _ = get_place_cells(
         session=session, spks=spks, rewarded=rewarded, config=config, plot=plot
     )
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/grosmark_analysis.py:76]: Whoops, might it be that the code prompting your comment went elsewhere due to me editing it?

File: viral/sequence_utils.py (hunk 19 above)
Comment (josefbitzenhofer) [reply viral/sequence_utils.py:114]: Hmm, do you think I should try and do that on the unsmoothed?

File: viral/sequence_utils.py (hunk 24 above)
Comment (josefbitzenhofer) [reply viral/sequence_utils.py:131]: So the MATLAB code is different depending on whether you construct this xy_by_bin for "linear" or "circular".
Or is your question if I tested it against the MATLAB code? I am 99% sure I did, but if necessary, I'd have to test again.

File: viral/sequence_utils.py (hunk 23 above)
Comment (josefbitzenhofer) [reply viral/sequence_utils.py:290]: so in the matlab code they randomly shuffle bisn within the event, but I haven't checked the outputs (in the Grosmark code, the permutation and significance check are done in two different places). Guess I should check it against MATLAB?

File: viral/sequence_utils.py (hunk 25 above)
Comment (josefbitzenhofer) [reply viral/sequence_utils.py:346]: I tested it and now we're golden
